### PR TITLE
move readers under metrics supervisor and require registered reader

### PR DESCRIPTION
**Pull request by tsloughter**
This patch moves readers under a readers supervisor instead of being
started from the metrics server. Readers must be declared in config
and provided with a unique atom name that is used for registration
of the process. This name is also used to create named ets tables
to store the view aggregation mappings and the metrics.

Use of named ets tables and registered processes means the pid and
tids don't need to be tracked but a future update may find it
better to remove the need for names.

### CODE REVIEW

**Comment by tsloughter** [comment]
Still not happy with it, but good enough to move forward with I think.

File: apps/opentelemetry/src/opentelemetry_sup.erl [hunk 1]
```diff
@@ -69,7 +69,7 @@ init([Opts]) ->
                 start => {otel_span_sup, start_link, [Opts]},
                 type => supervisor,
                 restart => permanent,
-                shutdown => 5000,
+                shutdown => infinity,
```
Comment (ferd) [inline apps/opentelemetry/src/opentelemetry_sup.erl:72]: Initially did a double-take about `infinity` but I realize the lower-level supervisors have a timeout, so transitively this is making sure the children supervisors have their time to shut down cleanly while timeouts still exist within the system.

File: apps/opentelemetry_experimental/test/otel_test_utils.hrl [hunk 2]
```diff
@@ -2,12 +2,16 @@
...
-                           case X of
+                           try X of
                                true ->
                                    ok;
                                false ->
                                    timer:sleep(100),
                                    Until(I+1)
+                           catch
+                               _:_ ->
+                                   timer:sleep(100),
```
Comment (ferd) [inline apps/opentelemetry_experimental/test/otel_test_utils.hrl:13]: should we ct:pal this so there's at least a trace of failures in logs if the case is unsuccessful?

**Review by ferd — APPROVED** [review]

File: apps/opentelemetry_experimental/test/otel_test_utils.hrl (hunk 2 above)
Comment (tsloughter) [reply apps/opentelemetry_experimental/test/otel_test_utils.hrl:13]: Ah, good point. I added it for the specific case where I expect an exception while the process is coming up, but will definitely be confusing later on when something else is throwing an exception.

File: apps/opentelemetry/src/opentelemetry_sup.erl (hunk 1 above)
Comment (tsloughter) [reply apps/opentelemetry/src/opentelemetry_sup.erl:72]: Only the workers at the lower level should be non-infinity, you said "lower-level" supervisors, did you find some that aren't infinity?

**Pull request merged**